### DISCUSSION TITLE
Dependencies not set within async tests

### DISCUSSION THREAD

**Discussion by nashysolutions** [pointfreeco/swift-dependencies#55] [Q&A]
When testing with the keyword async, dependencies fail to get set.

```swift
func testMe() async {
        let fetcher = withDependencies {
            $0.analytics = .mock
        } operation: {
            HomePayloadFetcher()
        }
        // ... etc
        // ... @Dependency(\.analytics) has no test implementation, but was accessed from a test context:
}
```

Removing the async keyword resolves the issue.

```swift
func testMe() {
        let fetcher = withDependencies {
            $0.analytics = .mock
        } operation: {
            HomePayloadFetcher()
        }
        // ... etc
        // ... works
}
```

I cannot show the code with the problem but it does fail consistently.

The attached project has a basic implementation of the issue I'm having. Unfortunately, the test has started passing for me, but it definitely did fail two or three times. If it's passing for you too, then could you maybe theorise as to why this code would fail?

[BugReport.zip](https://github.com/pointfreeco/swift-dependencies/files/10838909/BugReport.zip)

**Comment by iampatbrown** [comment]
This might be similar to #46 where the failure is actually being caused by the app host. See [Testing gotchas](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/testing/#Testing-gotchas) and the issue above for more info. Sorry on my phone atm so can't test your code.

**Reply by tgrapperon** [reply]
It doesn't seem to be that. You can remove the whole ViewModel from the View and it still fails. It works of you copy/paste the dependency into the test. I have to go, so I can't dive deeper ATM neither.

**Comment by tgrapperon** [comment]
@nashysolutions You're linking with the `Dependency` library in your app (this is good) and your tests target (this creates duplicate implementations). This can create mismatches with `KeyPath`s depending on which `DependencyValues` it'll use. 
Tests are passing if you remove the `Dependencies` dependency from your test target. Your test target will depend from `Dependencies` transitively, through the copy that is bundled with your app.
Someone had a similar issues a few days ago, so I think this should be documented somewhere. I don't know if we can do ObjC introspection to show a message. In any case Xcode shows warnings like:
```
objc[70954]: Class _TtC12DependenciesP33_9EC5E77E274EAECEB523307B89F04B0A12CachedValues is 
implemented in both /Users/Thomas/Library/Developer/XCTestDevices/575895D6-F32E-4777-A1F4-
0D6350378E13/data/Containers/Bundle/Application/4FEAC191-230B-45E1-B8A5-
80ECA7D2A0D6/BugReport.app/BugReport (0x10040d528) and 
/Users/Thomas/Library/Developer/XCTestDevices/575895D6-F32E-4777-A1F4-
0D6350378E13/data/Containers/Bundle/Application/4FEAC191-230B-45E1-B8A5-
80ECA7D2A0D6/BugReport.app/PlugIns/BugReportTests.xctest/BugReportTests (0x106ee9468). One of the 
two will be used. Which one is undefined.
```

**Reply by tgrapperon** [reply]
I've submitted #56 to document this issue.

**Reply by nashysolutions** [reply]
Hi thanks for letting me know. But this hasn't resolved my issue.

**Reply by tgrapperon** [reply]
This was the issue in your `BugReport` though. Removing the `Dependencies` dependency from the test target reliably fixes the issue.
Depending on your project's structure, it can be a similar issue, the host app accessing dependencies as @iampatbrown suggested, or maybe some escaping closure lurking somewhere, and escaping the `withDependencies`'s operation scope.

**Reply by nashysolutions** [reply]
Yes you're right: it was due to the host app running in parallel to the tests, as documented [here](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/testing/#Testing-gotchas).

Thanks!
